src/verifier/step4/SumcheckLogic.sol

```diff
@@ -0,0 +1,488 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "src/pasta/Pallas.sol";
+import "src/pasta/Vesta.sol";
+import "src/verifier/step4/KeccakTranscript.sol";
+import "src/verifier/step4/EqPolynomial.sol";
+
+library PallasPolyLib {
+    struct MLPoly {
+        uint64 numvars;
+        uint256[] Z;
+    }
+
+    struct SparseEntry {
+        uint256 idx;
+        uint256 entry;
+    }
+
+    struct SparsePoly {
+        uint64 numvars;
+        SparseEntry[] Z;
+    }
+
+    struct UniPoly {
+        uint256[] coeffs;
+    }
+
+    struct CompressedUniPoly {
+        uint256[] coeffs_except_linear_term;
+    }
+
+    function newML(uint256[] memory Z, uint64 numvars) public pure returns (MLPoly memory) {
+        require(Z.length == 2 ^ numvars);
+        return MLPoly(numvars, Z);
+    }
+
+    function boundPolyVarTop(MLPoly calldata P, uint256 r) public pure returns (MLPoly memory) {
+        uint256 mid = P.Z.length / 2;
+
+        uint256[] memory new_Z;
+
+        for (uint256 i = 0; i < mid; i++) {
+            uint256 a = P.Z[i];
+            uint256 b = P.Z[mid + i];
+            // new_Z[i] = a + r * (b - a);
+            new_Z[i] = addmod(a, mulmod(r, addmod(b, Pallas.negateBase(a), Pallas.P_MOD), Pallas.P_MOD), Pallas.P_MOD);
+        }
+
+        return MLPoly(P.numvars - 1, new_Z);
+    }
+
+    function evaluate(MLPoly calldata P, uint256[] calldata r) public pure returns (uint256 result) {
```

Comment:
We can make this function generic, by forcing caller to pass correct modulus as parameter